drivers/counter/counter_imx_epit.c

```diff
@@ -0,0 +1,187 @@
+}
+
+static void imx_epit_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	EPIT_Type *base = get_epit_config(dev)->base;
+	struct imx_epit_data *driver_data = dev->driver_data;
+
+	EPIT_ClearStatusFlag(base);
+
+	if (driver_data->callback != NULL) {
+		driver_data->callback(dev, (void *)driver_data->user_data);
+	}
+}
+
+static void imx_epit_init(struct device *dev)
+{
+	struct imx_epit_config *config = (struct imx_epit_config *)
+							   get_epit_config(dev);
+	EPIT_Type *base = config->base;
+	epit_init_config_t epit_config = {
+		.freeRun     = true,
+		.waitEnable  = true,
+		.stopEnable  = true,
+		.dbgEnable   = true,
+		.enableMode  = true
+	};
+
+	/* Adjust frequency in the counter configuration info */
+	config->info.freq = get_epit_clock_freq(base)/(config->prescaler + 1U);
+
+	EPIT_Init(base, &epit_config);
+}
+
+static int imx_epit_start(struct device *dev)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+
+	/* Set EPIT clock source */
+	EPIT_SetClockSource(base, epitClockSourcePeriph);
+
+	/* Set prescaler */
+	EPIT_SetPrescaler(base, get_epit_config(dev)->prescaler);
+
+	/* Start the counter */
+	EPIT_Enable(base);
+
+	return 0;
+}
+
+static int imx_epit_stop(struct device *dev)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+
+	/* Disable EPIT */
+	EPIT_Disable(base);
+
+	return 0;
+}
+
+static u32_t imx_epit_read(struct device *dev)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+	u32_t value;
+
+	value = EPIT_GetCounterLoadValue(base) - EPIT_ReadCounter(base);
+
+	return value;
+}
+
+static int imx_epit_set_top_value(struct device *dev, u32_t ticks,
+				  counter_top_callback_t callback,
+				  void *user_data)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+	struct imx_epit_data *driver_data = dev->driver_data;
+
+	/* Disable EPIT Output Compare interrupt for consistency */
+	EPIT_SetIntCmd(base, false);
+
+	driver_data->callback = callback;
+	driver_data->user_data = user_data;
+
+	/* Set both reload and counter values to "ticks" */
+	EPIT_SetOverwriteCounter(base, true);
+	EPIT_SetCounterLoadValue(base, ticks);
+
+	if (callback != NULL) {
+		/* (Re)enable EPIT Output Compare interrupt */
+		EPIT_SetIntCmd(base, true);
+	}
+
+	return 0;
+}
+
+static u32_t imx_epit_get_pending_int(struct device *dev)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+
+	return EPIT_GetStatusFlag(base) ? 1U : 0U;
+}
+
+static u32_t imx_epit_get_top_value(struct device *dev)
+{
+	EPIT_Type *base = get_epit_config(dev)->base;
+
+	return EPIT_GetCounterLoadValue(base);
+}
+
+static u32_t imx_epit_get_max_relative_alarm(struct device *dev)
+{
+	return COUNTER_MAX_RELOAD;
+}
+
+static const struct counter_driver_api imx_epit_driver_api = {
+	.start = imx_epit_start,
+	.stop = imx_epit_stop,
+	.read = imx_epit_read,
+	.set_top_value = imx_epit_set_top_value,
+	.get_pending_int = imx_epit_get_pending_int,
+	.get_top_value = imx_epit_get_top_value,
+	.get_max_relative_alarm = imx_epit_get_max_relative_alarm,
```

Comment:
@stanislav-poboril wondering why you implemented get_max_relative_alarm if alarm support isn't implemented in EPIT driver?

Reply:
I think the reason was that it is still valid to call `counter_set_channel_alarm()` when alarm support is not implemented (number of channels is zero) as it will return `-ENOTSUP` . While calling `counter_get_max_relative_alarm()` will try to invoke invalid function pointer and crash. On the other hand, why would anybody do it? The `get_max_relative_alarm` can be removed from EPIT counter driver if needed.